### ISSUE TITLE
Flyway: Vendor based SQL script selection

### ISSUE THREAD

**Issue by akumar074** [quarkusio/quarkus#30650]
### Description

Support for vendor placeholder which can select the SQL scripts from {vendor} location.
```
quarkus.flyway.locations=classpath:db/migration/{vendor}
```

Vendors can be automatically detected by flyway, and migration scripts will be executed accordingly.

Here's how it works in spring-boot - [spring docs](https://docs.spring.io/spring-boot/docs/2.1.1.RELEASE/reference/html/howto-database-initialization.html#howto-execute-flyway-database-migrations-on-startup).

**Comment by geoand** [comment]
This can certainly be done, but as it would make the code a little more complex, I'd like us to wait until we have more user demand before going forward.

**Comment by akumar074** [comment]
@geoand Thanks for the clarification. Can you please suggest a workaround for now to update the `quarkus.flyway.locations`, based on jdbc driver/dependecy provided. I tried setting up using `RunTimeConfigurationDefaultBuildItem`, but this only sets the configuration at runtime and Flyway doesn't allow this change.

**Comment by geoand** [comment]
@akumar074 `quarkus.flyway.locations` is a build time configuration property, it cannot be changed at runtime.
If you are looking to set the property from an extension, we would need to add a new build item to support that case.

**Comment by akumar074** [comment]
Is there any Configuration BuildItem for build time available already in quarkus?

**Comment by geoand** [comment]
No, but you can likely configure the Smallrye Config object used at build time using other means.
@radcortez do we have any examples of extensions adding build time ConfigSource ?

**Comment by radcortez** [comment]
It is possible with the ServiceLoader mechanism. Register any supported services in the `deployment` module, and they will be discovered and added to the build time Config.

**Comment by geoand** [comment]
Sure yeah, I was just looking for an example so users can just copy the pattern :)

**Comment by radcortez** [comment]
I don't think we have.

I recommend using a `io.smallrye.config.ConfigSourceFactory` to access the current configuration and create a new `ConfigSource` with a higher ordinal to override the values.

**Comment by akumar074** [comment]
Thanks, @geoand @radcortez, I'll give it a try.

**Comment by akumar074** [comment]
Hey guys, I tried the ConfigSource way, and It worked. However, the problem with this is that we are using `quarkus.datasource.db-kind` property to identify the db-kind and this is an optional property when we are having a single db. 
We found that it's possible to fetch the db-kind using `DefaultDataSourceDbKindBuildItem`. Can you please help us how we can set the ConfigSource using the Build item.
Thanks.

**Comment by geoand** [comment]
I actually suggest using `AggregatedDataSourceBuildTimeConfigBuildItem` which contains all the resolved information.

You would essentially need a build step that records the data using your own recorder (essentially it would stash the information in a static field of the recorder) and then your ConfigSourceFactory would just read the static field to get the info.

**Comment by geoand** [comment]
Yes, it has been around a very long time. It's part of `quarkus-agroal-deployment`

**Comment by akumar074** [comment]
Hey sorry, the message got deleted. Thanks a lot for quick help.

**Comment by geoand** [comment]
🙏🏼

**Comment by akumar074** [comment]
Hi, I tried to use Recorder to save the value from Processor. The ConfigSourceFactory always gets null in this. Here's the snippet from the code. I feel I am missing something little here. Please take a look.

Recorder:
```
@Recorder
public class MyRecorder {
    private static String dbKind;

    public static void setDbKind(String value) {
        dbKind = value;
    }

    public static String getDbKind() {
        return dbKind;
    }
}
```
ConfigSourceFactory:
```
@Override
    public Iterable<ConfigSource> getConfigSources(ConfigSourceContext context) {
        Map<String, String> configuration = new HashMap<>();
        final String databaseName = KogitoAddOnPersistenceJDBCRecorder.getDbKind();
        configuration.put(FLYWAY_LOCATIONS, LOCATION_PREFIX + getDBName(databaseName));
        return List.of(new KogitoAddOnPersistenceJDBCConfigSource(configuration));
    }
```
Processor:
```
@BuildStep
    @Record(ExecutionTime.STATIC_INIT)
    @Produce(ArtifactResultBuildItem.class)
    public void configure(List<DefaultDataSourceDbKindBuildItem> dbKind, KogitoAddOnPersistenceJDBCRecorder recorder) {
        recorder.setDbKind(dbKind.get(0).getDbKind());
    }
```

**Comment by geoand** [comment]
Can you try something like:

```java
    @BuildStep
    @Record(ExecutionTime.STATIC_INIT)
    @Produce(ArtifactResultBuildItem.class)
    public void configure(List<DefaultDataSourceDbKindBuildItem> dbKind, BuildProducer<RunTimeConfigBuilderBuildItem> runTimeConfigBuilder, KogitoAddOnPersistenceJDBCRecorder recorder) {
        recorder.setDbKind(dbKind.get(0).getDbKind());
    }
```

**Comment by akumar074** [comment]
I tried it, but It is still getting a null value.

**Comment by radcortez** [comment]
In this case, you may want to use `RunTimeConfigurationDefaultBuildItem`. This records a configuration name / value in the `RunTimeDefaultsConfigSource`, available in static init and runtime. The source has a lower ordinal, so the value can be easily overridden.

**Comment by akumar074** [comment]
Hey @radcortez , I can see `RunTimeConfigurationDefaultBuildItem` is in `quarkus-core-deployment` but I am unable to find `RunTimeDefaultsConfigSource`. Are you talking about `RuntimeConfigSource`?

**Comment by akumar074** [comment]
I just did that and we got `quarkus.flyway.locations is set to 'classpath:/db/postgresql' but it is build time fixed to 'db/migration'. Did you change the property quarkus.flyway.locations after building the application?`

**Comment by radcortez** [comment]
This is expected because we record the original configuration from build time and match it against any change of the same configuration at runtime. The goal is to notify the user that he may be changing a configuration value that will not have any effect because it builds time.

In this case, you must also use `SuppressNonRuntimeConfigChangedWarningBuildItem` to suppress the warning.

**Comment by akumar074** [comment]
Due to `quakrus.flyway.locations` being build time property. this doesn't override the property. we have just the above message.

**Comment by radcortez** [comment]
But at which phase are you looking to have the Flyway configuration? I thought you wanted the config at runtime. Can you please confirm?

**Comment by akumar074** [comment]
We want to override the property `quarkus.flyway.locations` at build time.

**Comment by radcortez** [comment]
Ops, I am so sorry. For some reason, I thought you wanted this available at runtime. In this case, the build steps we proposed will not help.

Ideally, you should resolve everything at the `ConfigSourceFactory` level. Once you get into the build steps execution, the `Config` is already initialized. If the issue is not being able to resolve the database vendor (because it is only resolved at the agroal extension level), then you have to provide a way to update the source from the factory, to add extra configuration.

The source instance can be obtained by calling `ConfigProvider.getConfig().unwrap(SmallRyeConfig.class).getConfigSource(NAME | Class)`.

**Comment by akumar074** [comment]
Thanks, @radcortez , and I am sorry I was not clear earlier regarding the change. I am able to change the ConfigSource as you mentioned earlier (Using ConfigSource and its factory). 
The challenge for us is how we can **find the Datasource type** and **then update the ConfigSource**. As you mentioned the **Build steps execution happens after the configs are initialized**, We tried the build steps to fetch the DB type, but then we could not update the config source.

**Comment by radcortez** [comment]
Yes, the trick is to provide a way to update the source. You can access your source by name or type with `ConfigProvider.getConfig().unwrap(SmallRyeConfig.class).getConfigSource(NAME | Class).` and then update it. I usually don't recommend this method, but if the configuration name being updated was not used yet, and you are sure your update is executed before the configuration is read, then it should be fine.

There is probably an additional issue. The config root class is populated at the config start, so it won't see your update. You would need to retrieve the configuration property manually.

I'm thinking of an alternative implementation. Instead of changing the configuration, you could enhance the Flyway extension to merge the Datasource type result and the Quarkys Flyway configuration to resolve the desired configuration.

**Comment by asm0dey** [comment]
At the same time, I want it at runtime, as Spring does — to be able to ship an application with a local database (h2 for example) and a remote (for example, PostgreSQL), and let users decide what they want
